### ISSUE TITLE
Calculation method for solar fractions in HOM

### ISSUE THREAD

**Issue by FWuellhorst** [RWTH-EBC/AixLib#918]
**What is the problem?**

In order to validate the HOM, we need a intuitiv and phyiscal modelling of the short wave solar radiation entering through a window.

**How do we want to solve it? Describe the solution you'd like**

We need to identify how we can use solar fractions and shape factors to calculate which wall receives a certain percentage of the radiation through a window.
Main goals are:

- [x] All calculations should be based on sources
- [x] The parametrization needs to be intuitiv
- [x] The results should be close to the results provided in the ASHRAE Standard

**Related sources**
- [Book on shape factors - Section 9.4](http://160592857366.free.fr/joe/ebooks/Mechanical%20Engineering%20Books%20Collection/HEAT%20TRANSFER/Ptinciples%20of%20Heat%20Transfer.pdf)
- ANSI/ASHRAE Standard 140-2017

**Problems with the current implementation**

Heres the list of all problems i identified during the process:
- Different approaches for multiple windows. In `ASHRAE140.EastWestFacingWindows`, a `MultiSum` is used, in `ASHRAE140.SouthFacingWindows`, no sum is used, as only one window is present. I would propose a central approach. 
- A `RealInput` is used for the short waved radiation. The 'Buildings' library for example has it's own port.
- Solar distribution factors are inside of each wall. However, the factors are calulated in the room they are applied in.

**Comment by FWuellhorst** [comment]
First Assumption:
Both ASHRAE and Buildings Library make the assumption that all short wave radiation from windows first hit the floor. Afterwards, the radiaton is reflected as diffuse radation onto the other areas.
This assumption make the modelling much easier, as:
- We don't have to care about whether the sun may hit a wall early in the morning or late in the eveing. Although this calculation would be possible in theory, the resulting parameters would be not constant but trajectories. Generated state events during simulation and thus longer simulation time don't match the minimal added detail.
- We also don't have to calculate special pathways of a lightbeam after it hits the floor.

I will therefore go ahead and also make the assumption for this model.

Second Assumption:
As we wan't to model according to ASHREA, i will stick with the "four bounces" described in the appendix. However, they are dealing with one ´alpha´ for all surfaces and make further assumptions. As we have all necessary values either way in our records, i will extend the approach to surface dependent factors.

**Comment by PMehrfeld** [comment]
Thanks @FWuellhorst.  
@konstantinaxanthopoulou: Any comments here? Agree, disagree?

**Comment by FWuellhorst** [comment]
@PMehrfeld and @konstantinaxanthopoulou :
Any particular reason why there is no `PartialRoom` model?
As all rooms have at least a floor, ceiling 3-4 walls and same interfaces, a parent model does make a lot of sense from my point of view. I could quickly add this if you'd like

**Comment by PMehrfeld** [comment]
@FWuellhorst: In branch `HOM_project` exists a partial room model (`AixLib.ThermalZones.HighOrder.Rooms.BaseClasses`). But I think currently the two ASHRAE room models do not extend from this partial room model.  The current partial room model does not include any wall, so that the actual room model is open for individual geometry and e.g. instance names of the wall parts.

You could definitely extend the partial room for the ASHRAE rooms. But discuss this with @konstantinaxanthopoulou.

For the validation test cases I already explained to @konstantinaxanthopoulou that creating the base case and extending from this one in subsequent test cases just makes sense. She will cover this.

**Comment by konstantinaxanthopoulou** [comment]
@PMehrfeld @FWuellhorst 
Thanks @FWuellhorst, this sounds good to me. 

I did extend the two ASHRAE Rooms and the validation test cases from `PartialRoom` (`AixLib.ThermalZones.HighOrder.Rooms.BaseClasses`) in the branch i am currently working on (#883) but the changes are not yet in `HOM_project`

**Comment by PMehrfeld** [comment]
@konstantinaxanthopoulou:

Would you please create a pull request for `issue883_newTestCases` --> `HOM_project`. And assign me as reviewer.

I am not going to delete the branch `issue883_newTestCases` afterwards. Thus, you can continue working in this branch. But we have the latest developments in `HOM_project`.

**Comment by konstantinaxanthopoulou** [comment]
@FWuellhorst 
`HOM_project` is now updated and includes the latest developments.

**Comment by FWuellhorst** [comment]
So, after a day of coding i've come up with a possible solution. The main problem when solving this issue is the nasty parametrization in my opinion.
Therefore, i made a new expandable connector `Utilities.Interfaces.ShortRadSurf` for short wave radiation used on surfaces. It's basically a bus connector, with the following variables:
```  
  Modelica.SIunits.Power QRad_out "Output short waved radiation from the room to the surface";
  Modelica.SIunits.Power QRad_in "Input short waved radiation from the surface to the room";

  Modelica.SIunits.Length L "Length of surface" annotation(HideResult=false);
  Modelica.SIunits.Height H "Height of surface" annotation(HideResult=false);
  Real eps "Emissivity of surface" annotation(HideResult=false);
  Real rho "Reflectivity of surface" annotation(HideResult=false);
  Real tau "Transmissivity of surface" annotation(HideResult=false);
  Real alpha "Absorptivity of surface, equal to eps or 1-rho-tau" annotation(HideResult=false);
```

The main advantage of this approach is that all parameters are automatically assign in the `wall` and `window` model:
![grafik](https://user-images.githubusercontent.com/22442514/81950448-39ae9600-9604-11ea-98af-61afcf173785.png)
Afterwards, you just specify the numbers of walls, ceilings etc. in the new model and connect the buses. 
The new model, `AixLib.Utilities.HeatTransfer.SolarRadInRoom` takes mutliple window connectors as input and output to walls, ceilings, floors and windows (latter may be relevant for internal windows, ASHRAE includes this).

Although all values are constant (except from the heat flow), the use of a connector makes it necessary to declare all parameters inside `AixLib.Utilities.HeatTransfer.SolarRadInRoom` as not constant.
To implement the approach from the ASHRAE standard, I (currently) had to make the following assumptions:

- All floors act as one joined area. However, the heat flows are seperated based on areas etc. at the end according to ASHRAE
-  All walls have the same heigh +  ceiling and floor have (joined together) the same area. This is necessary in order to use the approximations of sight factors from the standard (or the provided book). 
- The sight factors currently are independend of wall orientation. This i will implement in the next stage. However, a lot of new parameters are necessary to do this. For a square room however, the orientation has no influence.

I tested all newly implemented models in a test enviroment. However, when checking the room models i get a singular system cause of some connection to `airflow`. @PMehrfeld and @konstantinaxanthopoulou : Do you know the error? If not, i will check why this error occurs.
![grafik](https://user-images.githubusercontent.com/22442514/81951432-7038e080-9605-11ea-8677-724119824e10.png)

**Comment by FWuellhorst** [comment]
This is the working brach:
https://github.com/RWTH-EBC/AixLib/commits/issue918_HOMSolarFractions

**Comment by FWuellhorst** [comment]
Nevermind, i fixed the error.
The model now works. However, i still have to check the results against the table in ASHRAE.
Conversion scripts for the deleted parameters are done.

Before i test everything, what do you guys think of the approach. The need of constant Connectors is a bit nasty, however makes the whole setup pretty easy.

**Comment by PMehrfeld** [comment]
@FWuellhorst : Thanks for your effort. However, I do not understand, why you need a bus connector with variables that are actually parameters. I still think a model should make it as well?!

**Comment by FWuellhorst** [comment]
@PMehrfeld : The main problem is, that all wall parameters that we need for the calculations are inside the wall model, the window parameters even one layer deeper in the wall. Therefore, when we build a new room model, we would have to manually duplicate all wall and window parameteres and add it to the model. For the standard case of 4 walls, 1 ceiling etc. thats easy, but what for 5 walls and e.g. 2 floors. I could not think of a way to dynamically add the parameter specification.

Then i tried parameters in a connector, however the values are local for each connector.
As all calculations are done outside of `equation`, the fact that we use parameters should not affect model speed.

An other option would be a record array, however we still have to manually add them. The order of the array would be crucial, therefore the user would have to: 1. Make the connections and write down which array element is connected to which wall and 2. add the wall records in the correct order to the model. For this to work, we would also have to extend the wall record or make a new one, as we need: Length, Height, eps, rho... which are currently in different records.


Thats why i chose this approach. If you know a better approach, please let me know!

**Comment by PMehrfeld** [comment]
@FWuellhorst: Please proceed as discussed. Thanks for your modelling work.

**Comment by FWuellhorst** [comment]
## A problem we have:
The current `wall` model does not specify where the window is located in the wall. This geometric information would be necessary to calculate the correct sight factors for the windows. Furthermore, if we dont know where the windows are, we can't quantify the wall sight factor.
Assumptions about where the window is located would be heuristic and maybe difficult to follow. To make these parameters a requirement would decrease the usability of the HOM Model, as detailed construction plans would be necessary.
Also, all these calculations would only work if exaclty 4 walls are specified. The moment a user splits a wall into two walls (which may happen), the easy calculation of sight factors from floor to wall becomes impossible. For this case, we would also need a long list of heuristics and new parameters.

## Solution approach:
The one sight factor we can always calculate is from floor to ceiling (as long as both don't have windows).
As energy must be conserved,
```
sight_factor_floor_walls + sight_factor_floor_ceil + sight_factor_floor_win = 1
```
must hold.
Based on the the assumption **that the radation is evenly distributed among walls and windows for the second bounce**, we can obtain:
```
sight_factor_floor_win = A_win / (A_win + A_walls) * (1 - sight_factor_floor_ceil)
```
and 

```
sight_factor_floor_walls = A_walls / (A_win + A_walls) * (1 - sight_factor_floor_ceil)
```
Then, the energy is further distributed to each element using:
```
sight_factor_floor_wall_i = A_wall_i / A_walls * sight_factor_floor_walls 
```
For the windows it would be the same.

Main advantages:
- We can still use the current `wall` and `window` models.
- We don't have to specify where a window is or deal with the placements of different walls
- We guarrante a correct energy balance. Using the functions, numerical errors could lead to the sum of sight factors being different from 1
- The influence on the results is relativly small. The floor typically reflects only 10 % of the incoming radiation. Approx. 50 % go to the ceiling, the rest to the walls. Effects like correct window placement etc. on the temperatures of the walls will therefore be of a small magnitude.

**Comment by FWuellhorst** [comment]
@konstantinaxanthopoulou : Do you know which Testcase or Room is used to obtain the results given in Table B7-2 in the Appendix of the standard? I am comparing results and wanted to check if the parameters of the room are all properly set.

**Comment by konstantinaxanthopoulou** [comment]
@FWuellhorst : Yes of course. The South Window Cases of Table B7-2 are Case270, Case290, Case320. The East/West Window Cases are Case 300 and Case 310. Do you need any additional information about the these Cases?

**Comment by PMehrfeld** [comment]
@FWuellhorst : Thanks for your detailed description!

Especially, the last bullet of your advantages is important to conserve here online and in the model documentation as well.

**Comment by FWuellhorst** [comment]
I documented the assumptions in the model and checked if the mentioned test cases produce the correct values.
For floor and windows, the values (almost) match (0.9024 instead of 0.903 and 0.0062 instead of 0.006).
The ceiling however has a big gap: 0.0511 instead of 0.039. I checked all formulas but could not track down the error. According to the sight factor equation and the plot, the sight factor for the room is around 0.54. If we enter this value into the second bounce equation, the solar fraction for the ceiling should be at least 0.0486. 
As the ceiling also affects the walls with our current assumptions, this gap further leads to different wall sight factors.
@konstantinaxanthopoulou : Any idea on why the value for the ceiling is so low? Did the ASHREA exactly follow their equations? Or maybe different room geometries?

**Comment by konstantinaxanthopoulou** [comment]
@FWuellhorst : I agree the value for the ceiling should be much higher. The room geometries don't change at all between the cases. It is always the same rectangle 8x6x2.7 and only the window orientation changes.
I will have a look at this again and come back to you soon.

**Comment by FWuellhorst** [comment]
- [ ] Add inner wall parameter to docstring to make clear the parameters are for the inner wall surface
- [ ] Document the difference of our implementation and the tables in the ASHREA Standard. We could not find out why their sight factor to the ceiling is so low.

I will do the remaining tasks together with the simulate and plot scripts for the test cases.

**Comment by FWuellhorst** [comment]
- [ ] Make manual choice possible to conform to standard
- [ ] Add simulate and plot scripts

**Comment by FWuellhorst** [comment]
@konstantinaxanthopoulou : I added a little "know limitations" to the record you created for the EastWest Room. I could not find the values for abs=0.1 and i think you took the ones from the south window case, or?

Furthermore i think it's strange that the tables in the Appendix do not list any "window absorbed" radiation, as the calculation regard this solar fractions. Do they have alpha=0 for the window? They mentioned that alpha is valid for ALL surfaces. Makes no sense to me...

**Comment by konstantinaxanthopoulou** [comment]
@FWuellhorst : Thanks for the addition. Yes exactly, there are no values for abs=0.1 for the EastWest Room so i took the ones from the other one.

Regarding the window absorbed radiation, do you mean the solar radiation that is initially absorbed when the radiation passes through the window? The Standard generally declares that fractional values for the walls with windows (i.e., the south wall) include the portion of the solar radiation absorbed by the glass (as it passes back
out the window) and conducted back into the zone. Solar radiation absorbed by the glass (and conducted inward) as it initially passes into the building is not included in the values.

**Comment by FWuellhorst** [comment]
I mean the radiation that initialy hits the floor and then it transmitted through the window into the ambient again.
So in the tables, the fraction for the window is the part which is lost to the ambient and the part which is absorbed is added to the wall where the window is located?
At the end it makes no difference for our results, as we neither model the loss to the ambient nor the absorbtion of the window.

**Comment by konstantinaxanthopoulou** [comment]
Yes exactly, that's how I understand it too.

**Comment by MartinRaetz** [comment]
@konstantinaxanthopoulou @FWuellhorst 
Is there a status update?
Should this Issue remain active?

**Comment by FWuellhorst** [comment]
@PMehrfeld could also know the current status

**Comment by PMehrfeld** [comment]
Yes, it's on my radar that I have to review PR #965.  
Thanks for the reminder.

**Comment by PMehrfeld** [comment]
Closed by PR #965
